fractal_tasks_core/tasks/_registration_utils.py

```diff
@@ -0,0 +1,247 @@
+# Copyright 2024 (C) BioVisionCenter
+#
+# Original authors:
+# Joel Lüthi <joel.luethi@uzh.ch>
+#
+# This file is part of Fractal
+"""Utils functions for registration"""
+import copy
+
+import anndata as ad
+import numpy as np
+import pandas as pd
+import zarr
+
+from fractal_tasks_core.ngff.specs import Well
+
+
+def _split_well_path_image_path(zarr_url: str):
+    """
+    Returns path to well folder for HCS OME-Zarr zarr_url
+    """
+    zarr_url = zarr_url.rstrip("/")
+    well_path = "/".join(zarr_url.split("/")[:-1])
+    img_path = zarr_url.split("/")[-1]
+    return well_path, img_path
+
+
+def create_well_acquisition_dict(
+    zarr_urls: list[str],
+) -> dict[str, dict[int, str]]:
+    """
+    Parses zarr_urls & groups them by HCS wells & acquisition
+
+    Generates a dict with keys a unique description of the acquisition
+    (e.g. plate + well for HCS plates). The values are a dictionary. The keys
+    of the secondary dictionary are the acqusitions, its values the zarr_url
+    for a given acquisition.
+
+    Args:
+        zarr_urls: List of zarr_urls
+
+    Returns:
+        image_groups
+    """
+    image_groups = dict()
+
+    # Dict to cache well-level metadata
+    well_metadata = dict()
+    for zarr_url in zarr_urls:
+        well_path, img_sub_path = _split_well_path_image_path(zarr_url)
+        # For the first zarr_url of a well, load the well metadata and
+        # initialize the image_groups dict
+        if well_path not in image_groups:
+            image_groups[well_path] = {}
+            well_group = zarr.open_group(well_path, mode="r")
+            well_metadata[well_path] = Well(**well_group.attrs.asdict())
```

Comment:
I'm confused by the use of the `Well` model here.

Let's say that `well_path="/somewhere/plate.zarr/B/03/`. Then I expect `well_group.attrs.asdict()` to have a single key, named "well". The value associated to this key is the one that fits with the `Well` model (i.e. it has `images` and `version` attributes).

Therefore I guess we should use `NgffWellMeta` here, rather than `Well. Even better, we should just use something like
```python
well_metadata[well_path] = load_NgffWellMeta(well_path).well
```

I'll try to proceed this way, and see whether something goes wrong.

Note that _if_ my review is correct then this change would likely not have any effect, because the current behavior would always have `images = []`, and then we would never enter in the `for image in ...` loop.

(ref #693)

Reply:
I think this was actually working due to a glitch somewhere else.
The 3D well group has this .zattrs file
```json
{
    "well": {
        "images": [
            {
                "acquisition": 0,
                "path": "0"
            },
            {
                "acquisition": 1,
                "path": "1"
            }
        ],
        "version": "0.4"
    }
}
```
while the 2D one has this one
```json
{
    "images": [
        {
            "acquisition": 0,
            "path": "0"
        },
        {
            "acquisition": 1,
            "path": "1"
        }
    ],
    "version": "0.4"
}
```

Reply:
I fixed this both in copy-ome-zarr (bffe65aff13671abb133db4e96a46e4bff26ef53) and then in this registration-related function (f413291426ec0ec623dd5de9e27081f3bc8543dc).

Reply:
@jluethi This is now clear on my side, let me know if it makes sense to you.

Reply:
Ah, I was also confused on why `NgffWellMeta` wasn't working here, but that explains it! Thanks for the fix!